File: Data Structures & Algorithms/count-number-of-islands/submission-7.java
class Solution {
    int ROWS,COLS;
    public int numIslands(char[][] grid) {
        ROWS = grid.length;
        COLS = grid[0].length;
        int res = 0;
        for(int i=0;i<ROWS;i++){
            for(int j=0;j<COLS;j++){
                if(grid[i][j]=='1'){
                    res++;
                    count(i,j,grid);
                }
            }
        }
        return res;
        
        
    }

    public void count(int i,int j, char[][] grid){
        if(i<0 || j<0 || i==ROWS || j == COLS || grid[i][j] != '1'){
            return;
        }
        grid[i][j] = '0';
        count(i+1,j,grid);
        count(i,j+1,grid);
        count(i,j-1,grid);
        count(i-1,j,grid);

    }
}

File: Data Structures & Algorithms/swim-in-rising-water/submission-7.py
class Solution:
    def swimInWater(self, grid: List[List[int]]) -> int:
        heap = [[grid[0][0],0,0]]
        res = 0
        M,N = len(grid),len(grid[0])
        vis = set()
        dir = [[0,1],[1,0],[0,-1],[-1,0]]
        vis.add((0,0))
        while heap:
            level,i,j = heapq.heappop(heap)
            res = max(res,level)
            if i == M-1 and j ==N-1:
                return res
            
            for dx,dy in dir:
                nx,ny = i+dx,j+dy
                if (nx,ny) in vis or nx<0 or nx ==M or ny<0 or ny ==N:
                    continue
                vis.add((nx,ny))
                heapq.heappush(heap,[grid[nx][ny],nx,ny])
        








        